### PR TITLE
feat: add custom roles

### LINKED ISSUE #13911: Add the ability to create custom roles in the UI
Users should have the ability to create, edit, and delete custom roles within our UI.

- Create custom roles MUI combobox for selecting existing custom roles and adding new custom roles
- Display full list of resources, actions, descriptions for a custom role using a checkbox list
- Display hardcoded subset of resources, actions, descriptions by default
- Create button to toggle between full resource list and subset resource list
- Mutate selected resources/actions when creating/editing a custom role
- Delete custom role

### PR DESCRIPTION
**Pull request by jaaydenh**
Resolve #13911

- [x] Design custom roles table when empty
- [x] Setup custom table loader skeleton for roles table
- [x] Hide custom roles tab if experiment is not enabled
- [x] Display a message that the name field cannot be modified
- [x] When editing a custom, the name field should not be editable
- [x] Create permission check for resource assign_org_role with action create
- [x] Set the checked state for permissions checkboxes to the current resource and actions for a role
- [x] Edit existing custom role page
- [x] Create new custom role page
- [x] Retrieve organization roles and display in a table
- [x] EditRolesButton falls back to role name when no display name is present
- [x] Add cancel and save buttons to top of create/edit role form
- [x] Display a subset of resources, add a checkbox to display all resources (audit_log, group, template, organization_member, provisioner_daemon, workspace)
- [x] Storybook stories

Screenshot 2024-08-02 at 2 58 56 PM

Screenshot 2024-08-02 at 3 00 03 PM

Screenshot 2024-08-02 at 3 01 56 PM

### CODE REVIEW

File: site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePage.tsx
Comment (aslilac) [inline site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePage.tsx:97]: You have this `PageHeader` here without an actual header (just actions), and then you have another `PageHeader` in the view component that contains the actual header without any actions. It should all just be one `PageHeader`.

File: site/src/pages/ManagementSettingsPage/CustomRolesPage/CustomRolesPage.tsx
Comment (aslilac) [inline site/src/pages/ManagementSettingsPage/CustomRolesPage/CustomRolesPage.tsx:21]:
```suggestion
  const { organization } = useParams() as { organization: string };
```

This page is only accessible from routes with an `:organization` param.

File: site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx
Comment (aslilac) [inline site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx:160]:
```suggestion
                            &ndash;{" "}
```

super nit picky, but an n-dash feels more appropriate here than a hyphen.

File: site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx
Comment (aslilac) [inline site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx:154]: maybe this (or at least part of it) could be pulled out into a variable with a name that helps explain all this

**Review by aslilac — COMMENTED** [review]
couple smalls things that could be cleaned up before merging, but overall looks great! 😄

File: site/src/pages/ManagementSettingsPage/CustomRolesPage/CustomRolesPage.tsx
Comment (jaaydenh) [reply site/src/pages/ManagementSettingsPage/CustomRolesPage/CustomRolesPage.tsx:21]: Steven mentioned today that custom roles should be independent from organizations and still accessible if organizations is not enabled

File: site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx
Comment (jaaydenh) [reply site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx:160]: got it, btw, this was meant to be temporary as I wanted to create a cleaner separation of the action and action description.

File: site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx
Comment (aslilac) [inline site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx:123]: this makes me sad, but not in any way that's really your fault 😂

File: site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx
Comment (aslilac) [inline site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx:158]: unnecessary fragment?

File: site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx
Comment (aslilac) [inline site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx:189]:
```suggestion
            {Object.entries(resourceActions).map(([resourceKey, value]) => (
```

nit: could avoid a level of indention by using single expression/implicit return syntax

File: site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx
Comment (aslilac) [inline site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx:208]:
```suggestion
                                    )
```

it should accept `undefined` as a `false`ish value right?

File: site/src/pages/ManagementSettingsPage/CustomRolesPage/CustomRolesPageView.tsx
Comment (aslilac) [inline site/src/pages/ManagementSettingsPage/CustomRolesPage/CustomRolesPageView.tsx:36]:
```suggestion
  const isLoading = roles === undefined;
```

this expression is already a boolean

alternatively, you could just do `!roles`

File: site/src/contexts/auth/permissions.tsx
Comment (aslilac) [inline site/src/contexts/auth/permissions.tsx:140]: I'm worried about this check. it doesn't specify a specific org or `any_org`, so I'm not sure what response to expect here, and it looks like the check that you added to the `organizationPermissions` query is correct, and is the permission value that I think we should actually be using for all of this.

cc @Emyrk

**Review by aslilac — COMMENTED** [review]
my main concern is the `assign_org_role` check in `AuthProvider`. I think we should remove that and just use the check in the `organizationPermissions` query everywhere. I tagged Steven so he can give a second opinion, because I'm not super familiar with the permissions checks stuff.

other than that, just some minor things! looking good :)

File: site/src/contexts/auth/permissions.tsx
Comment (Emyrk) [reply site/src/contexts/auth/permissions.tsx:140]: Yea this does not seem right. We should remove this imo, I see the other check as well

**Comment by BrunoQuaresma** [comment]
I will rely on @aslilac review for the FE code and focus on the design and QA.

1. If a form does not have more than one section, it should be a simple vertical form. I also would add a subtitle to the header.

Form

2. For the checkbox group, what do you think about having something like what we are using for [notifications](https://github.com/coder/coder/pull/14175)? Of course, replace it with the permissions primary and secondary text, and remove the icon on the right.

Form (1)

**Review by BrunoQuaresma — COMMENTED** [review]
I was not able to QA it.  I started a Coder server using `CODER_EXPERIMENTS=*,multi-organization,custom_roles ./scripts/develop.sh`,  added a [premium license](https://www.notion.so/coderhq/Licensing-7ff2140f2ba44158b57c78d0c30cd8b4?pvs=4#84f6567b7a2a43f59f62fbd44baac712), but I still not seeing the"roles" item in the sidebar for my org. 
Screenshot 2024-08-07 at 10 21 48
Screenshot 2024-08-07 at 10 21 42

File: site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx
Comment (jaaydenh) [reply site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx:208]: yes, this is probably leftover from an earlier iteration

File: site/src/contexts/auth/permissions.tsx
Comment (jaaydenh) [reply site/src/contexts/auth/permissions.tsx:140]: Got it, I originally was trying to get custom roles working independently from organizations

**Comment by aslilac** [comment]
@BrunoQuaresma you've got an underscore in there, not a hyphen

**Comment by jaaydenh** [comment]
@BrunoQuaresma Regarding using switches for organization roles similar to notifications.

1. FYI, I have been using the github example below as a reference. I feel like roles are conceptually different to notifications because roles are a set of things a users does or does not possess, whereas notifications feel like something that is turned on/off. I also did a quick survey of other tools and found that Azure, Github and many others use check boxes so my feeling is that it is a more common design pattern for managing custom roles.

2. I am guessing that an API call is made every time a switch is changed for notifications. For roles, I imagined that the user would go through the flow of checking all the relevant roles and submitting the changes because it felt like creating a set of things as opposed to switching individual settings on/off. Although, I do think there could be value in making each change in the checkbox send the patch request instead of waiting to press the submit button if others agree.

github

**Comment by jaaydenh** [comment]
@BrunoQuaresma regarding the form only having one section, I still need to add the delete role functionality and was thinking of doing it like organization settings. Unless, you have any thoughts about a better place to delete roles?

Screenshot 2024-08-07 at 3 35 58 PM

**Comment by BrunoQuaresma** [comment]
Related to the checkboxes: I think using GitHub as a reference is a good idea, but I missed a few things like the alignment of the descriptions and the "header checkbox" to check or uncheck the children options.

Related to the delete section: I don't think we need it on the create page right? Also, the width page for those settings pages is smaller because the sidebar does not give proper space to a horizontal form. Even using two sections, we can still use vertical forms. GH does this very well on their settings pages.

I will leave those decisions up to you.

**Comment by BrunoQuaresma** [comment]
While QAing it I recorded a video to share a few design things that could be improved.

https://github.com/user-attachments/assets/a2adca95-cddc-4cf7-8a00-8d934ffbee64

**Comment by jaaydenh** [comment]
@aslilac I made all the changes planned for this PR based on the suggestions. Design improvements for checkboxes and delete custom role will be completed in a separate PR.

File: site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx
Comment (aslilac) [inline site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx:232]: I love the addition of this! could we add something similar to the bottom of the table too, just to make sure people can find it? a small extra row that just says "show all permissions" like this does

File: site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx
Comment (aslilac) [reply site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx:222]: marking as unresolved again because I still see this in the code :p

**Review by aslilac — APPROVED** [review]

File: site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx
Comment (jaaydenh) [reply site/src/pages/ManagementSettingsPage/CustomRolesPage/CreateEditRolePageView.tsx:232]: I added the show all permissions checkbox in a table footer row

**Pull request merged**
